core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java

```diff
@@ -20,9 +20,19 @@
 package org.elasticsearch.common.inject;
 
 /**
+ * This interface can be added to a Module to spawn sub modules. DO NOT USE.
  *
+ * This is fundamentally broken.
+ * <ul>
+ * <li>If you have a plugin with multiple modules, return all the modules at once.</li>
+ * <li>If you are trying to make the implementation of a module "pluggable", don't do it.
+ * This is not extendable because custom implementations (using onModule) cannot be
+ * registered before spawnModules() is called.</li>
+ * </ul>
+ * @deprecated Create modules directly
  */
-public interface SpawnModules {
+@Deprecated
```

Comment:
I think we should reduce the scope here, remove the deprecation and open another issue to remove `SpawnModules` altogether. Elasticsearch core has gone through major refactorings to reduce the massive over-abstraction and over-engineering that we know as legacy code. Every project will get there and following the path we went down the last couple of weeks we should really make thinks clear and simple. Simple here means we need to add clean extension points without any modules depending on modules etc. you can extend a module and that's it. We should flatten out these extension points and don't allow for this recursive abstractions. It's hard enough to support one layer, multiple layers just make things harder and harder.